Phase 2 - Getting Started in Tech doc first draft (#469)
codebar/tutorials: general/getting-started-in-tech.md

---
layout: page
title: Getting Started in Tech
---

> Technology is a vast subject. It can enable you to write a to-do list on your phone, sell goods online or send a space probe to Neptune. Software is built by developers and codebar is here to teach you how to write such software. Whether you want to make a career of it or use your skills on personal projects.

Reviewer: ```suggestion
Technology is a vast subject. It can enable you to write a to-do list on your phone, sell goods online or send a space probe to Neptune. Software is built by developers and codebar is here to teach you how to write such software, whether you want to make a career of it or acquire some exta skills.
```
I think we should avoid the user of the personal projects wording, given it's something we tend to have problems with at workshops!

Author: very true 👍

Typically software can fall into two (very broad) categories:

- **User interface** - How to present information to people using the technology and enable them to interact with it. For example:
    - Displaying your vehicle and responding to joystick inputs in a game of Mario Kart
    - Displaying your balance and allowing you to make a payment in your bank’s online banking site
    - Displaying current temperature and allowing you to change the target on your Hive or Nest smart thermostat
- **Data processing** - How to manipulate the data and perform the task that the software implements. For example:
    - Calculating the position of your vehicle and detecting collisions
    - Validating that your account has enough funds to send this payment and processing the payment
    - Starting or stopping your heating or air conditioning depending on the current temperature and the target you set

At codebar, we concentrate on web technologies for several reasons:

> - No initial investment is needed: all the tools you need to learn about web technologies available online for free. And the computer you use doesn’t have to be new or a top of the range model, most computers will perform well, including shared computers in libraries.

Reviewer 2: technologies available → technologies **are** available

> - It doesn't need any initial investment: all the tools you need to learn about web technologies available online for free. And the computer you use doesn’t have to be new or a top of the range model, most computers will perform well, including shared computers in libraries.

Reviewer: Small update, just because we refer to multiple web technologies but then use singular.
```suggestion
- No initial investment is needed: all the tools you need to learn about web technologies available online for free. And the computer you use doesn’t have to be new or a top of the range model; most computers will perform well, including shared computers in libraries.
```

- You can see results very fast.
- They are ubiquitous: a lot of technologies that started on the web are now used in many other places. Even if you end up working in a domain that has nothing to do with the web, you will likely use or interact with web technologies.

Web technologies are designed in such a way that they tend to more clearly separate user interface work from data processing work. As a result, web developer jobs typically fall into three broad categories:
- Front-end: work that deals with user interfaces
- Back-end: work that deals with data processing

> - Full stack: work that combines both aspects

Reviewer 2: I would _not_ hyphenate any of the three areas. [People](https://english.stackexchange.com/questions/484817/full-stack-vs-full-stack-back-end-vs-back-end-front-end-vs-front-end) seem to hyphenate when they are adjectives (“_the front-end code_”) and not hyphenate when they are nouns (“_I work on the front end_”).

ADD GRAPHIC

To build a website like this involves two pieces of code:
- The front-end code that runs in the web browser
- The back-end code that runs on the web server

Below we got into more depth about these roles, and what languages you would need to learn.

## Front-end Development

As a front-end developer you would be responsible for the code that runs in the web browser and displays information to visitors. It is typically the practice of writing HTML, CSS and JavaScript.

#### HTML (Hyper Text Markup Language)

HTML is the language used to build websites. All text and content that you see on the internet is built using HTML.

[HTML Tutorials]()

#### CSS (Cascading Style Sheets)

CSS is used alongside HTML to style the page and make it look good.

[CSS Tutorials]()

#### JavaScript

JavaScript is used to change what is displayed on a web page in response to user activity.

[JavaScript Tutorials]()

## Back-end Development

> As a back-end developer you would be responsible for the code that runs on the server. This type of development usually consists of three parts: a server, an application, and a database. Many back end developers know front end languages such as HTML and CSS but need to use languages such as Ruby on Rails, Python, PHP, Java, and .Net to get the job done.

Reviewer 2: back end developers → back-end developers

Reviewer 2: front end languages → front-end languages

Reviewer 2: .Net → .NET

> These two roles overlap a fair amount so starting in one role doesn’t close the other one should you decide to change later. They also have strong ties to other areas of technology. For instance, front-end development shares a lot with mobile app development; back-end development with data science.

Reviewer: ```suggestion
Front-end and Back-end development overlap a fair amount so starting in one role doesn’t mean that you can't switch or learn the other, should you change your mind further down the line. They also have strong ties to other areas of technology. For instance, front-end development shares a lot with mobile app development; back-end development with data science.
```

#### Ruby
LINE ABOUT RUBY

[Ruby Tutorials]()

#### Python

> Python is the preferred option for data science thanks to scientific libraries like numpy and pandas.

Reviewer 2: numpy → NumPy

[Python Tutorials]()

#### Node.js
Node.js is an open-source, cross-platform, back-end JavaScript runtime environment that runs on the V8 engine and executes JavaScript code outside a web browser.

[Node Tutorials]()

#### PHP
LINE ABOUT PHP

[PHP Tutorials]()

### Databases

#### SQL
LINE ABOUT SQL

[SQP Tutorials]()

#### No-SQL
LINE ABOUT No-SQL

[Tutorial Link]()

> ## Full stack Development

Reviewer 2: Full stack → Full-stack

> A full stack web developer is a person who can develop both client (front-end) and server (back-end) software. For this we suggest learning any of the languages under front-end development and back-end development. Below are some further subjects that relate to both:

Reviewer 2: full stack → full-stack

---
layout: page
title: Getting Started in Tech
---

> Technology is a vast subject. It can enable you to write a to-do list on your phone, sell goods online or send a space probe to Neptune. Software is built by developers and codebar is here to teach you how to write such software. Whether you want to make a career of it or acquire some exta skills.

Reviewer 3: To keep it topical, maybe we could replace "send a space probe to Neptune" with "send a rover and a helicopter to Mars".

    - Calculating the position of your vehicle and detecting collisions
    - Validating that your account has enough funds to send this payment and processing the payment
    - Starting or stopping your heating or air conditioning depending on the current temperature and the target you set

At codebar, we concentrate on web technologies for several reasons:

- No initial investment is needed: all the tools you need to learn about web technologies are available online for free. And the computer you use doesn’t have to be new or a top of the range model, most computers will perform well, including shared computers in libraries.
- You can see results very fast.
- They are ubiquitous: a lot of technologies that started on the web are now used in many other places. Even if you end up working in a domain that has nothing to do with the web, you will likely use or interact with web technologies.

Web technologies are designed in such a way that they tend to more clearly separate user interface work from data processing work. As a result, web developer jobs typically fall into three broad categories:

> - Front end: work that deals with user interfaces

Reviewer 4: @KimberleyCook do you need a space before this line for the bullet points to display properly?

- Back end: work that deals with data processing
- Full stack: work that combines both aspects

Both front-end and back-end have their own complexities and neither is easier than the other. So let’s have a look at what they both involve. Take this web page as an example. It is a page on the codebar.io website and you are reading it via a web browser on a computer or mobile device. The codebar website runs on a computer in a data center that runs a piece of software called a web server. Both talk to each other via the internet.

To build a website like this involves two pieces of code:
- The front-end code that runs in the web browser
- The back-end code that runs on the web server

> Below we got into more depth about these roles, and what languages you would need to learn.

Reviewer 4: ```suggestion
Below we go into more depth about these roles, and what languages you would need to learn.
```

Author: nice catch!

## Front-end Development

As a front-end developer you would be responsible for the code that runs in the web browser and displays information to visitors. It is typically the practice of writing HTML, CSS and JavaScript.

> #### HTML (Hyper Text Markup Language)

Reviewer 3: Should that be a level 3 heading considering the previous one is level 2?

HTML is the language used to build websites. All text and content that you see on the internet is built using HTML.

> #### CSS (Cascading Style Sheets)

Reviewer 3: Level 3 heading?

CSS is used alongside HTML to style the page and make it look good.

> 

Reviewer 4: ```suggestion
```

@KimberleyCook cheeky extra line here.

> #### JavaScript

Reviewer 3: Level 3 heading?

JavaScript is used to change what is displayed on a web page in response to user activity.

## Back-end Development

> As a back-end developer you would be responsible for the code that runs on the server. This type of development usually consists of three parts: a server, an application, and a database. Many back-end developers know front-end languages such as HTML and CSS but need to use languages such as Ruby on Rails, Python, PHP, Java, and .NET to get the job done.

Reviewer 4: @KimberleyCook I just noticed this but I think here we should say Ruby instead of Ruby on Rails. Ruby is the language and Rails is the framework built on top of it.

Front-end and Back-end development overlap a fair amount so starting in one role doesn’t mean that you can't switch or learn the other, should you change your mind further down the line. They also have strong ties to other areas of technology. For instance, front-end development shares a lot with mobile app development; back-end development with data science.

> #### Ruby

Reviewer 3: Level 3 heading?

> Ruby is an object oriented scripting programming language typically used for building web applications. The most common implementation of Ruby is Rails web, the development framework built with Ruby.

Reviewer 4: @KimberleyCook I think we should use the full name of the framework here, e.g. Ruby on Rails. How about rewording this sentence like this?

```suggestion
Ruby is an object oriented scripting programming language typically used for building web applications. It is most commonly used with Ruby on Rails, a web development framework built with Ruby.
```

> #### Python

Reviewer 3: Level 3 heading?

Python is the preferred option for data science thanks to scientific libraries like NumPy and pandas.

> #### Node.js

Reviewer 3: Level 3 heading?

> Node.js is an open-source, cross-platform, back-end JavaScript runtime environment that runs on the V8 engine and executes JavaScript code outside a web browser.

Reviewer 3: The thing about the V8 engine is a bit of a geeky definition. What about? "Node.js is a development framework built with JavaScript. It allows you to write back-end components with the same technologies as the ones used in front-end work, which makes it particularly attractive when the same developers work on the front-end and back-end."

> #### PHP

Reviewer 3: Level 3 heading?

PHP is a general-purpose scripting language especially suited to web development. It is a server side scripting language that is embedded in HTML. It is used to manage dynamic content, databases, session tracking, even build entire e-commerce sites. Also commonly used with WordPress so if custom WordPress themes interests you, then learn PHP.

> #### SQL

Reviewer 3: Level 3 heading?

> SQL is a domain-specific language designed for managing data held in a relational database management system.

Reviewer 3: Add: "It is often combined with the technologies described above."

## Full-stack Development

A full-stack web developer is a person who can develop both client (front-end) and server (back-end) software. For this we suggest learning any of the languages under front-end development and back-end development. Below are some further subjects that relate to both:

> #### Testing

Reviewer 3: Level 3 heading?

Why do you want to test and what different types of tests you can do:
- Unit testing
- Integration testing
- End to end testing

> - Acceptance testing: Cucumber/Gherkin + shared understanding with business

Reviewer 4: @KimberleyCook if we are adding libraries + examples here shouldn't we add it to the other entries as well?

Author: So just say `acceptance testing` ??

Reviewer 4: @KimberleyCook yeah. I think we either add examples or a small paragraph explaining what they are to all of them or none of them, it feels weird that this kind of testing is singled out. For a newbie probably none of the other terms will make sense either.

- Security testing
- Performance testing

> #### REST API

Reviewer 3: Level 3 heading?

> 

Reviewer 4: ```suggestion
```

@KimberleyCook another extra cheeky line.

Basics on HTTP, JSON and how to put this together to use or build an API.

> #### Privacy

Reviewer 3: Level 3 heading?

Basics on privacy, things like GDPR and overall principles.

### Accessibility

> We feel this is particularly important if you're wanting to be a front-end developer.

Reviewer 3: Suggested wording:

Web technologies were originally designed to disseminate information to as wide a public as possible. As a result, they are very flexible and over the years the number and variety of devices able to access the web has exploded while underlying technologies like HTML have remained relatively stable. How you use web technologies to build your websites has an impact on who can access and interact with the information you provide. Accessibility is a set of guidance and techniques to ensure that your websites can be used by as wide a variety of people as possible, irrespective of whether they get on the web like you do or in a very different way.

Author: This feels very wordy 🤔  ???

I think we can skip the first bit and say `How you use web technologies to build your websites has an impact on who can access and interact with the information you provide. Accessibility is a set of guidelines and techniques to ensure that your websites can be used by as wide a variety of people as possible, irrespective of whether they get on the web like you do or in a very different way.`

Reviewer 3: Works for me.
